Add recipe for org-upcoming-modeline

unhammer opened the pull request:
Show next org event in mode line

### Brief summary of what the package does

Shows the next upcoming org-mode event in your mode-line. You can click it to go to the event, or right-click for a menu of goto/snooze/clock in.

### Direct link to the package repository

https://github.com/unhammer/org-upcoming-modeline

### Your association with the package

Maintainer / responsible for all bugs.

### Relevant communications with the upstream package maintainer

None needed.

### Checklist


- [X] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [X] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [X] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [X] My elisp byte-compiles cleanly
- [X] I've used `M-x checkdoc` to check the package's documentation strings
- [X] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

riscy commented:
Thanks for this.  Quick first pass --

`org-upcoming-modeline.el` with byte-compile using Emacs 28.1:
```
org-upcoming-modeline.el:35:11: Warning: docstring wider than 80 characters
org-upcoming-modeline.el:39:57: Warning: Unused lexical variable `second'
In org-upcoming-modeline--set-string:
org-upcoming-modeline.el:133:8: Warning: docstring wider than 80 characters
org-upcoming-modeline.el:164:31: Warning: reference to free variable
    `org-upcoming-modeline-map'
org-upcoming-modeline.el:227:22: Warning: reference to free variable
    `ts-upcoming'
org-upcoming-modeline.el:252:11: Warning: reference to free variable `from'
org-upcoming-modeline.el:252:16: Warning: reference to free variable `to'
org-upcoming-modeline.el:252:19: Warning: reference to free variable `_on'
org-upcoming-modeline.el:252:23: Warning: reference to free variable `regexp'
org-upcoming-modeline.el:270:56: Warning: reference to free variable `limit'
org-upcoming-modeline.el:262:80: Warning: reference to free variable
    `match-group'
org-upcoming-modeline.el:273:52: Warning: Unused lexical variable `regexp'
In org-upcoming-modeline-snooze:
org-upcoming-modeline.el:344:4: Warning: Unused lexical argument `event'
In end of data:
org-upcoming-modeline.el:270:63: Warning: the function `limit' is not known to
    be defined.
org-upcoming-modeline.el:252:31: Warning: the function `match-group' is not
    known to be defined.
org-upcoming-modeline.el:228:8: Warning: the function `&key' is not known to
    be defined.
org-upcoming-modeline.el:225:16: Warning: the function `org-ql--defpred' is
    not known to be defined.
```
Emacs attempts to expand/evaluate macros before it gets to anything else.  As a result I believe many(most) of these arethe result of a [wrong time](https://www.gnu.org/software/emacs/manual/html_node/elisp/Problems-with-Macros.html#Problems-with-Macros) macroexpansion.


`org-upcoming-modeline.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- org-upcoming-modeline.el#L291: It's safer to sharp-quote function names; use `#'`
```


Package and license:
- Add a LICENSE file to the repository

unhammer commented:
Oh, for some reason the `*Compile-Log*` buffer gets buried in my regular emacs session, so I thought it compiled cleanly, sorry about that. 

That melpazoid warning seems like a false positive? `'repeat` is just a truthy value to make the timer repeat, not a function.

I've dropped support for org-ql <=0.5 which let me remove the call to the old macro. I do get a compile warning about `org-show-entry` now being called `org-fold-show-entry` but you didn't so that probably appeared in newer org-mode versions, other than that the compile log now seems clean. Also added LICENSE.

riscy commented:
> That melpazoid warning seems like a false positive?

That's correct and I apologize for that noise.  Something for me to correct!

Here's the latest --

`org-upcoming-modeline.el` with byte-compile using Emacs 28.1:
```
org-upcoming-modeline.el:167:31: Warning: reference to free variable
    `org-upcoming-modeline-map'
```

I think this is because `org-upcoming-modeline-map` needs to be defined before being consumed by the macro in `org-upcoming-modeline--set-string`.

riscy commented:
(I also wonder why this is a `defconst` instead of a `defvar`?)

unhammer commented:
Changed to defvar and moved over first use site. I guess I need to do these byte-compiles in `emacs -Q` to consistently get the same warnings.

riscy commented:
Thank you!  Merging.
